### PR TITLE
Add benchmark for Nest, remove extra alloc in BenchmarkMarshalerField

### PR DESCRIPTION
**Pull request by prashantv**
Some pre-work for #87

### CODE REVIEW

File: json_encoder_bench_test.go [hunk 1]
```diff
@@ -34,6 +34,18 @@ type logRecord struct {
...
+	f := func(kv KeyValue) error {
+		kv.AddString("s", "string")
+		return nil
+	}
+	for i := 0; i < b.N; i++ {
+		enc := newJSONEncoder()
+		enc.Nest("nested", f)
+		enc.Free()
+	}
+}
```
Comment (akshayjshah) [inline json_encoder_bench_test.go]: Does this benchmark reflect how we expect this method to be used? I'd expected most users to pass an anonymous function, like this:

```go
enc := newJSONEncoder()
enc.Nest("nested", func(kv KeyValue) error {
  kv.AddString("s", "string")
  return nil
})
enc.Free()
```

File: logger_bench_test.go [hunk 2]
```diff
@@ -41,7 +41,7 @@ func (u user) MarshalLog(kv zap.KeyValue) error {
 	return nil
 }
 
-var _jane = user{
+var _jane = &user{
```
Comment (akshayjshah) [inline logger_bench_test.go]: Thanks for explaining this more to me offline. Can we also change the user struct's `MarshalLog` method to have a pointer receiver?

File: json_encoder_bench_test.go (hunk 1 above)
Comment (prashantv) [inline json_encoder_bench_test.go]: Good point, replaced with something that is more realistic by having an inline closure + use of a variable as the parameter (which actually makes it ~10ns slower).

File: logger_bench_test.go (hunk 2 above)
Comment (prashantv) [inline logger_bench_test.go]: 👍

**Comment by prashantv** [comment]
Updated

**Comment by akshayjshah** [comment]
lgtm. Merge away.

**Pull request merged**